src/wallet/wallet.h

```diff
@@ -509,8 +523,11 @@ class CWalletTx : public CMerkleTx
         READWRITE(nTimeReceived);
         READWRITE(fFromMe);
         READWRITE(fSpent);
-        // TODO:
-        //READWRITE(mapSaplingNoteData);
+
+        if (fOverwintered && nVersion >= SAPLING_TX_VERSION) 
+        {
+            READWRITE(mapSaplingNoteData);                
+        }
```

Comment:
Did we decide what to do about records written by 2.0.0 after Sapling activation, and then read by 2.0.1? Has that case been smoke-tested?

Reply:
There was some discussion around this, but I don't recall the final outcome. This has not been smoke tested as of 10/2/2018

Reply:
Based on looking at the `LoadWallet` logic in `src/wallet/walletdb.cpp`, what @mdr0id and I concluded was:

- The Sapling v4 transactions written by 2.0.0 would be read by 2.0.1, and the `READWRITE` above would throw an error due to insufficient data.
- This gets caught by the loading logic, where there is a special case for transaction read fails that schedules a from-genesis rescan.
- The rescan will load the transactions directly into memory from the block data, and then immediately write them to disk.
- As the transaction ID is used as the database key, the 2.0.1 write overwrites the 2.0.0 record, fixing it for the next time the node starts.

We have not tested this AFAIK.

Reply:
Need to sync 2.0.0 with testnet, tap the testnet faucet to get some funds (it will be v4 transaction), shut down, and launch node with this PR

Reply:
I can confirm the rescan logic is activated as we initially discussed and later concluded.

I synced up a wallet with testnet using v2.0.0 which had a v4 Sapling transaction (transparent value transfer) in the wallet.

Switching to 2.0.1 (this PR) triggered a rescan.

Switching back to 2.0.0 did not trigger a rescan (as expected, since the mapSaplingNoteData is appended to the end of the serialized data expected by 2.0.0)

Switching back to 2.0.1 (this PR) triggers a rescan.

When triggering the rescan, `debug.log` output looks like this:

```
2018-10-06 02:05:50 init message: Loading wallet...
2018-10-06 02:05:50 Warning: Warning: error reading wallet.dat! All keys read correctly, but transaction data or ad
dress book entries might be missing or incorrect.
2018-10-06 02:05:50  wallet                   31ms
2018-10-06 02:05:50 init message: Rescanning...
2018-10-06 02:05:50 Rescanning last 299187 blocks (from block 0)...
2018-10-06 02:05:50 AddToWallet a2f121c7219ff17b68de66963b4bb5220e372687076067d90717e4d670c2e2ee  new
...
...
2018-10-06 02:05:55 AddToWallet dda3941da3ab77f3bdc61728354e6e9285c31e456f38b4cc39278ff5c3e7d0fd  new
2018-10-06 02:06:00  rescan                 9629ms
```